Fix #707 - Error if class named 'await' in async function

ooblegork opened the pull request:
This adds the functionality of the following behavior: 

*  If class is named 'await' within an async function, report error.

There is also an added diagnostic type for the error.

ooblegork commented:
I have read and hereby agree to quick-lint-js' Contributor License Agreement (CLA-v1.md).

strager commented on src/parse-statement.cpp:
```diff
@@ -1130,7 +1130,11 @@ void parser::parse_and_visit_class_heading(
   case token_type::kw_yield:
     // TODO(#707): Disallow classes named 'await' in async functions.
-    // TODO(#707): Disallow classes named 'yield' in generator functions.
+    if (this->in_async_function_) {
+      this->diag_reporter_->report(
+          diag_cannot_declare_class_named_await_in_async_function{
+              .name = this->peek().identifier_name().span()});
+    }
```
heh, I didn't think it'd be this easy!

strager reviewed (changes requested):
Must fix: Add some tests. (Or did I miss your tests? It's hard to read your patch with all the formatting changes. @_@)

ooblegork commented on src/parse.cpp:
```diff
@@ -101,8 +101,8 @@ parser::binary_expression_builder::binary_expression_builder(
-expression* parser::binary_expression_builder::last_expression() const
-    noexcept {
+expression* parser::binary_expression_builder::last_expression()
+    const noexcept {
```
Going to fix this. I accidentally used 10.0. Do you think the difference in versions is what caused the excessive formatting changes?

strager replied:
Yes, definitely.

strager commented on src/parse-statement.cpp:
```diff
@@ -1130,7 +1130,11 @@ void parser::parse_and_visit_class_heading(
   case token_type::kw_await:
   case token_type::kw_yield:
     // TODO(#707): Disallow classes named 'await' in async functions.
-    // TODO(#707): Disallow classes named 'yield' in generator functions.
```
I'd leave the 'yield' comment alone, or replace `TODO(#707)` with `TODO(strager)`.

ooblegork replied:
I left the comment as-is. Once this issue is resolved, I plan to work on the class 'yield' behavior next.

strager commented on test/test-parse-function.cpp:
```diff
@@ -294,6 +294,33 @@ TEST(test_parse, await_in_async_function) {
+    EXPECT_THAT(v.errors, IsEmpty());
+  }
+
+  {
+    padded_string code(u8"async function g() {\nclass await {}\n}"_sv);
+    spy_visitor v;
+    parser p(&code, &v);
+    p.parse_and_visit_module(v);
+    EXPECT_THAT(
+        v.errors,
+        ElementsAre(DIAG_TYPE_OFFSETS(
+            &code, diag_cannot_declare_class_named_await_in_async_function,
+            name, strlen(u8"async function g() {\nclass "), u8"await")));
+  }
+}
```
Nit: I think the code is more readable with ` ` instead of `\n`:
```suggestion
  {
    padded_string code(u8"async function g() { class await {} }"_sv);
    spy_visitor v;
    parser p(&code, &v);
    p.parse_and_visit_module(v);
    EXPECT_THAT(
        v.errors,
        ElementsAre(DIAG_TYPE_OFFSETS(
            &code, diag_cannot_declare_class_named_await_in_async_function,
            name, strlen(u8"async function g() { class "), u8"await")));
  }
}
```

strager reviewed (approved):
Your patch looks great to me. Let me know when I should merge it.

ooblegork commented:
> Your patch looks great to me. Let me know when I should merge it.

Great. I fixed the small changes (re: moved tests into test-parse-class.cpp, removed '\n' characters). Feel free to merge whenever.

strager commented:
Landed in commit da5ea9eced559e092b087f7bdfc440999a894be5.
